File: packages/sdk/examples/governed-action.ts
/**
 * Governed action example — simulate, propose, approve via /v1/actions.
 *
 * Requires a running Commander API with Action Gateway enabled.
 *
 *   COMMANDER_API_URL=http://127.0.0.1:4000 COMMANDER_API_KEY=... \
 *     pnpm exec tsx packages/sdk/examples/governed-action.ts
 */

import { CommanderGatewayClient } from '../src/v1/client.js';

const baseUrl = process.env.COMMANDER_API_URL ?? 'http://127.0.0.1:4000';
const apiKey = process.env.COMMANDER_API_KEY;

if (!apiKey) {
  console.error('Set COMMANDER_API_KEY to call /v1/actions');
  process.exit(2);
}

const client = new CommanderGatewayClient({ baseUrl, apiKey });

const actionInput = {
  source: 'sdk-example',
  package: 'demo.package',
  model: 'demo-model',
  tool: 'ticket.create',
  destination: 'demo://tickets/approval',
  effectType: 'demo.ticket.create',
  args: { title: 'Reset password' },
  idempotencyKey: `sdk-example-${Date.now()}`,
};

const { simulation } = await client.simulateAction(actionInput);
console.log('simulate', simulation.effect, simulation.actionDigest);

const { action, accepted } = await client.proposeAction(actionInput);
console.log('propose', action.state, accepted);

if (action.decision.effect === 'require_approval') {
  const approved = await client.approveAction(action.runId, {
    actionDigest: simulation.actionDigest,
    simulationId: simulation.simulationId,
    policySnapshotId: simulation.policySnapshotId,
  });
  console.log('approve', approved.state);
}

const evidence = await client.getActionEvidence(action.runId);
console.log('evidence', evidence.verification);

File: packages/python-sdk/examples/governed_action.py
"""Governed action example — simulate, propose, approve via /v1/actions."""

from __future__ import annotations

import asyncio
import os
import time

from commander._gateway_client import CommanderGatewayClient
from commander._types import ActionApprovalInput, ProposeActionInput


async def main() -> None:
    api_key = os.environ.get("COMMANDER_API_KEY")
    if not api_key:
        raise SystemExit("Set COMMANDER_API_KEY to call /v1/actions")

    base_url = os.environ.get("COMMANDER_API_URL", "http://127.0.0.1:4000")
    action_input = ProposeActionInput(
        source="python-sdk-example",
        package="demo.package",
        model="demo-model",
        tool="ticket.create",
        destination="demo://tickets/approval",
        effect_type="demo.ticket.create",
        args={"title": "Reset password"},
        idempotency_key=f"python-example-{int(time.time())}",
    )

    async with CommanderGatewayClient(base_url=base_url, api_key=api_key) as client:
        simulation = await client.simulate_action(action_input)
        print("simulate", simulation.effect, simulation.action_digest)

        action, _replay, accepted = await client.propose_action(action_input)
        print("propose", action.state, accepted)

        if action.decision.effect == "require_approval":
            approved = await client.approve_action(
                action.run_id,
                ActionApprovalInput(
                    action_digest=simulation.action_digest,
                    simulation_id=simulation.simulation_id,
                    policy_snapshot_id=simulation.policy_snapshot_id,
                ),
            )
            print("approve", approved.state)

        evidence = await client.get_action_evidence(action.run_id)
        print("evidence", evidence.verification)


if __name__ == "__main__":
    asyncio.run(main())
